src/uct/ib/mlx5/dv/ib_mlx5dv_md.c

```diff
@@ -142,13 +156,95 @@ static ucs_status_t uct_ib_mlx5_devx_reg_atomic_key(uct_ib_md_t *ibmd,
+    UCT_IB_MLX5DV_SET64(mkc, mkc, start_addr, addr);
     UCT_IB_MLX5DV_SET64(mkc, mkc, len, length);
     UCT_IB_MLX5DV_SET(create_mkey_in, in, translations_octword_actual_size, list_size);
 
+    mr = mlx5dv_devx_obj_create(md->super.dev.ibv_context, in,
+                                uct_ib_mlx5_calc_mkey_inlen(list_size),
+                                out, sizeof(out));
+    if (mr == NULL) {
+        ucs_debug("mlx5dv_devx_obj_create(CREATE_MKEY, mode=KSM) failed, syndrome %x: %m",
+                  UCT_IB_MLX5DV_GET(create_mkey_out, out, syndrome));
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    *mr_p = mr;
+    *mkey = (UCT_IB_MLX5DV_GET(create_mkey_out, out, mkey_index) << 8) |
+            (addr & 0xff);
+
+    return UCS_OK;
+}
+
+static ucs_status_t
+uct_ib_mlx5_devx_reg_ksm_data(uct_ib_mlx5_md_t *md,
+                              uct_ib_mlx5_ksm_data_t *ksm_data,
+                              size_t length, off_t off,
+                              struct mlx5dv_devx_obj **mr_p,
+                              uint32_t *mkey)
+{
+    ucs_status_t status;
+    uint32_t *in;
+    void *klm;
+    int i;
+
+    status = uct_ib_mlx5_alloc_mkey_inbox(ksm_data->mr_num, &in);
+    if (status != UCS_OK) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    klm = UCT_IB_MLX5DV_ADDR_OF(create_mkey_in, in, klm_pas_mtt);
+    for (i = 0; i < ksm_data->mr_num; i++) {
+        UCT_IB_MLX5DV_SET64(klm, klm, address, (intptr_t)ksm_data->mrs[i]->addr);
+        UCT_IB_MLX5DV_SET(klm, klm, byte_count, ksm_data->mrs[i]->length);
+        UCT_IB_MLX5DV_SET(klm, klm, mkey, ksm_data->mrs[i]->lkey);
+        klm = UCS_PTR_BYTE_OFFSET(klm, UCT_IB_MLX5DV_ST_SZ_BYTES(klm));
+    }
+
+    status = uct_ib_mlx5_devx_reg_ksm(md, (intptr_t)ksm_data->mrs[0]->addr + off,
+                                      length, ksm_data->mr_num,
+                                      ksm_data->mrs[0]->length, in, mr_p, mkey);
+    ucs_free(in);
+    return status;
+}
+
+static ucs_status_t uct_ib_mlx5_devx_reg_atomic_key(uct_ib_md_t *ibmd,
+                                                    uct_ib_mem_t *ib_memh)
+{
+    uct_ib_mlx5_mem_t *memh = ucs_derived_of(ib_memh, uct_ib_mlx5_mem_t);
+    uct_ib_mlx5_md_t *md = ucs_derived_of(ibmd, uct_ib_mlx5_md_t);
+    off_t offset = uct_ib_md_atomic_offset(uct_ib_mlx5_md_get_atomic_mr_id(md));
+    struct ibv_mr *mr = memh->mr;
+    ucs_status_t status = UCS_OK;
+    size_t reg_length, length;
+    int list_size, i;
+    void *klm;
+    uint32_t *in;
+    intptr_t addr;
+
+    if (!(md->flags & UCT_IB_MLX5_MD_FLAG_KSM)) {
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    if (memh->super.flags & UCT_IB_MEM_MULTITHREADED) {
+        return uct_ib_mlx5_devx_reg_ksm_data(md, memh->ksm_data, memh->mr->length,
+                                             offset, &memh->ksm_data->atomic_dvmr,
+                                             &memh->super.atomic_rkey);
+    }
+
+    reg_length = UCT_IB_MD_MAX_MR_SIZE;
```

Comment:
do we really need the local var reg_length

Reply:
readability